packages/entity/src/EntityMutationTrigger.ts

```diff
@@ -0,0 +1,77 @@
+  TSelectedFields extends keyof TFields = keyof TFields
+> {
+  /**
+   * Trigger set that runs within the transaction but before the entity is created in the database.
+   */
+  beforeCreate?: EntityMutationTrigger<TFields, TID, TViewerContext, TEntity, TSelectedFields>[];
+  /**
+   * Trigger set that runs within the transaction but after the entity is created in the database and cache is invalidated.
+   */
+  afterCreate?: EntityMutationTrigger<TFields, TID, TViewerContext, TEntity, TSelectedFields>[];
+
+  /**
+   * Trigger set that runs within the transaction but before the entity is updated in the database.
+   */
+  beforeUpdate?: EntityMutationTrigger<TFields, TID, TViewerContext, TEntity, TSelectedFields>[];
+  /**
+   * Trigger set that runs within the transaction but after the entity is updated in the database and cache is invalidated.
+   */
+  afterUpdate?: EntityMutationTrigger<TFields, TID, TViewerContext, TEntity, TSelectedFields>[];
+
+  /**
+   * Trigger set that runs within the transaction but before the entity is deleted from the database.
+   */
+  beforeDelete?: EntityMutationTrigger<TFields, TID, TViewerContext, TEntity, TSelectedFields>[];
+  /**
+   * Trigger set that runs within the transaction but after the entity is deleted from the database and cache is invalidated.
+   */
+  afterDelete?: EntityMutationTrigger<TFields, TID, TViewerContext, TEntity, TSelectedFields>[];
+
+  /**
+   * Trigger set that runs within the transaction but before the entity is created, updated, or deleted.
+   */
+  beforeAll?: EntityMutationTrigger<TFields, TID, TViewerContext, TEntity, TSelectedFields>[];
+  /**
+   * Trigger set that runs within the transaction but before the entity is created in, updated in, or deleted from
+   * the database and the cache is invalidated.
+   */
+  afterAll?: EntityMutationTrigger<TFields, TID, TViewerContext, TEntity, TSelectedFields>[];
+
+  /**
+   * Trigger set that runs after committing the transaction unless one is supplied
+   * after any mutation (create, update, delete). If the call to the mutation is wrapped in a transaction,
+   * this too will be within the transaction.
+   */
+  afterCommit?: EntityMutationTrigger<TFields, TID, TViewerContext, TEntity, TSelectedFields>[];
+}
+
+/**
+ * A trigger is a way to specify entity mutation operation side-effects that run within the
+ * same transaction as the mutation itself. The one exception is afterCommit, which will run within
+ * the transaction if a transaction is supplied.
+ */
+export abstract class EntityMutationTrigger<
+  TFields,
+  TID,
+  TViewerContext extends ViewerContext,
+  TEntity extends ReadonlyEntity<TFields, TID, TViewerContext, TSelectedFields>,
+  TSelectedFields extends keyof TFields = keyof TFields
+> {
+  abstract async executeAsync(
+    viewerContext: TViewerContext,
+    queryContext: EntityQueryContext,
+    entity: TEntity
+  ): Promise<void>;
+}
```

Comment:
Unless there's a meaningful reason to make this a class, plain async functions would be more lightweight:
```ts
type EntityMutationTrigger<
    TFields,
    TID,
    TViewerContext extends ViewerContext,
    TEntity extends ReadonlyEntity<TFields, TID, TViewerContext, TSelectedFields>,
    TSelectedFields extends keyof TFields = keyof TFields
> = (
      viewerContext: TViewerContext,
      queryContext: EntityQueryContext,
      entity: TEntity
    ) => Promise<void>;
```

and the replacement for the constructor looks like:

```ts
function createThrowConditionallyTrigger<...>(parameters): EntityMutationTrigger<...> {
  return async function throwConditionallyAsync() {
    ... close over parameters if needed
  }
}
```

Reply:
So far the class approach has worked out fairly cleanly for `EntityPrivacyPolicyRule`, especially since feeding params into the constructor is common as is having overloaded constructors. I can imagine that these may turn out to be reasonably similar, but I'll definitely keep it in mind as we start writing triggers and seeing usage patterns.